### ISSUE TITLE
Integrate webpush notifications

### ISSUE THREAD

**Issue by arcturus** [glowing-bear/glowing-bear#941]
Following the comments after article:
https://ardeenelinfierno.com/2017/05/05/replacing-irccloud-with-open-web-technologies/

Looks like a good addition to the project. I'm pretty happy to contribute it and have it landed in master \o/.

But would like to have some guidance about how you imagine this feature included. In the top of my mind there are some points to talk about:

- Enabling webpush from the client: Right now I added a dirty hack to show some of the configuration strings in the settings screen. This can be hidden, just with a button once the user accepts to use push notifications (browser will ask). But also it will require for you to have the plugin installed, so don't know if that can be detected from glowing-bear, or is something that can be enabled/disabled via config.
- Need better integration with google cloude messaging service: This should not be a problem, as every deployment should have it's own GMC id in the manifest.json file, and this file can be read dinamycally, so the value will be fetch automatically too.
- Plugin should support multiple clients at the same time to be able to receive notifications in multiple devices. This is definitely an enhancement and will probably live in the plugin side, so yeah, kind of transparent for glowing-bear.

@lorenzhs @txanatan thoughts? I think actually the first point is the one that requires clarification, the rest is just a list of improvements to add to have a good quality.

Cheers!

**Comment by lorenzhs** [comment]
Thanks @arcturus! I'm not very familiar with how WebPush works but a quick look at the docs you linked sounds promising.

- We can read settings over the relay protocol, like this: https://github.com/glowing-bear/glowing-bear/blob/6af2a870fb0c09d934ae7c1aa5a10552955ea9ed/js/connection.js#L211 https://github.com/glowing-bear/glowing-bear/blob/6af2a870fb0c09d934ae7c1aa5a10552955ea9ed/js/connection.js#L446
- I'm not sure whether there's a nice way to *change* settings. Of course we could just send `/set` commands to WeeChat if not. We could offer setting everything up for the user, including sending `/script install` etc, but we should make clear that commands are being executed, and I'd like to offer a "show me how to do this manually" option. This would make it seamless for most users and give easy instructions to those who'd rather not have us execute commands for them.
- GMC IDs don't appear to be very secret so we could just register for one and put it in the manifest. People who host GB themselves can change it if they want.
- Of course Apple [requires a developer license](https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/NotificationProgrammingGuideForWebsites/PushNotifications/PushNotifications.html) to get a Push ID. Muppets. So no push notifications for iOS and macOS, I guess?
- It only works with HTTPS, which is another good reason to push everyone onto a secure connection. :)

**Comment by arcturus** [comment]
Hi,

yes, webpush is not a complicated technology, but requires some crytpo stuff to make it work.

The weechat plugin uses a webservice to proxy the crypto functions:
https://github.com/arcturus/webpushproxy

That proxy can be as well deployed in your own environment if user wants to own all the elements in the chain.

Some comments below:

> We can read settings over the relay protocol, like this: https://github.com/glowing-bear/glowing-bear/blob/6af2a870fb0c09d934ae7c1aa5a10552955ea9ed/js/connection.js#L211 https://github.com/glowing-bear/glowing-bear/blob/6af2a870fb0c09d934ae7c1aa5a10552955ea9ed/js/connection.js#L446

Thanks for the tip will look into this.

> I'm not sure whether there's a nice way to change settings. Of course we could just send /set commands to WeeChat if not. We could offer setting everything up for the user, including sending /script install etc, but we should make clear that commands are being executed, and I'd like to offer a "show me how to do this manually" option. This would make it seamless for most users and give easy instructions to those who'd rather not have us execute commands for them.

I can work on a first version that just output the list of commands, and later on iterate to see if there is a better UI experience.

> GMC IDs don't appear to be very secret so we could just register for one and put it in the manifest. People who host GB themselves can change it if they want.

Exactly, the version hosted in https://glowing-bear.org will need it's own, but then every deployment  (host) will require it's own, it's free and pretty easy to get one.

> Of course Apple requires a developer license to get a Push ID. Muppets. So no push notifications for iOS and macOS, I guess?

Unfortunately serviceworkers are not supported in ios yet (any browser as all of them are safari flavours), but any other browser in macosx (opera, chrome or firefox) will work. Fortunately safari guys have showed some interest in implementing serviceworkers, but that could take a long time to go.

> It only works with HTTPS, which is another good reason to push everyone onto a secure connection. :)

\o/ yeah, this is a plus!!

Thanks for the tips!

**Comment by lorenzhs** [comment]
Cool! It's too bad Safari doesn't support it, so iOS users will have to continue using pushover or something like that.

It seems that there is a [python version of the webpush crypto stuff](https://github.com/web-push-libs/pywebpush), so maybe that could be integrated into the weechat script, and no proxy would be required? Would make setup a little bit more complicated but :man_shrugging:

**Comment by arcturus** [comment]
Yeah, it's an option, but as I'm quite new on weechat scripting, didn't want to add more dependencies. But definitely we can hide that behind the weechat script.

Will open an issue to consider using the python library in the weechat script repo.

**Comment by olepbr** [comment]
Apologies if this isn't the right place/you'd like this to be a separate issue, but support for Web Push is in the newest iOS/iPadOS betas, which would allow glowing-hear to do push notifications on those platforms: https://webkit.org/blog/13878/web-push-for-web-apps-on-ios-and-ipados/
